File: src/extensibility/dotnet-extensibility/DataModel/ResourceCatalogBuilder.cs
using System.Text.Json;

namespace Nexus.DataModel
{
    /// <summary>
    /// A resource catalog builder simplifies building a resource catalog.
    /// </summary>
    public record ResourceCatalogBuilder
    {
        #region Fields

        private readonly string _id;
        private Dictionary<string, JsonElement>? _properties;
        private List<Resource>? _resources;

        #endregion

        #region Constructors

        /// <summary>
        /// Initializes a new instance of the <see cref="ResourceCatalogBuilder"/>.
        /// </summary>
        /// <param name="id">The identifier of the resource catalog to be built.</param>
        public ResourceCatalogBuilder(string id)
        {
            _id = id;
        }

        #endregion

        #region "Methods"

        /// <summary>
        /// Adds a property.
        /// </summary>
        /// <param name="key">The key of the property.</param>
        /// <param name="value">The value of the property.</param>
        /// <returns>The resource catalog builder.</returns>
        public ResourceCatalogBuilder WithProperty(string key, JsonElement value)
        {
            _properties ??= new();

            _properties[key] = value;

            return this;
        }

        /// <summary>
        /// Adds a property.
        /// </summary>
        /// <param name="key">The key of the property.</param>
        /// <param name="value">The value of the property.</param>
        /// <returns>The resource catalog builder.</returns>
        public ResourceCatalogBuilder WithProperty(string key, object value)
        {
            _properties ??= new();

            _properties[key] = JsonSerializer.SerializeToElement(value);

            return this;
        }

        /// <summary>
        /// Adds a <see cref="Resource"/>.
        /// </summary>
        /// <param name="resource">The <see cref="Resource"/>.</param>
        /// <returns>The resource catalog builder.</returns>
        public ResourceCatalogBuilder AddResource(Resource resource)
        {
            _resources ??= new List<Resource>();

            _resources.Add(resource);

            return this;
        }

        /// <summary>
        /// Adds a list of <see cref="Resource"/>.
        /// </summary>
        /// <param name="resources">The list of <see cref="Resource"/>.</param>
        /// <returns>The resource catalog builder.</returns>
        public ResourceCatalogBuilder AddResources(params Resource[] resources)
        {
            return AddResources((IEnumerable<Resource>)resources);
        }

        /// <summary>
        /// Adds a list of <see cref="Resource"/>.
        /// </summary>
        /// <param name="resources">The list of <see cref="Resource"/>.</param>
        /// <returns>The resource catalog builder.</returns>
        public ResourceCatalogBuilder AddResources(IEnumerable<Resource> resources)
        {
            _resources ??= new List<Resource>();

            _resources.AddRange(resources);

            return this;
        }

        /// <summary>
        /// Builds the <see cref="ResourceCatalog"/>.
        /// </summary>
        /// <returns>The <see cref="ResourceCatalog"/>.</returns>
        public ResourceCatalog Build()
        {
            return new ResourceCatalog(_id, _properties, _resources);
        }

        #endregion
    }
}

File: src/extensibility/rust-extensibility/src/data_model/resource_catalog_builder.rs
use std::collections::HashMap;

use super::{
    Resource, ResourceCatalog,
    resource_catalog::{ResourceCatalogId, Resources, ResourcesError},
};

const README: &str = "readme";
const LICENSE: &str = "license";

/// A resource catalog builder simplifies building a resource catalog.
pub struct ResourceCatalogBuilder {
    id: ResourceCatalogId,
    properties: Option<HashMap<String, String>>,
    resources: Option<Vec<Resource>>,
}

impl ResourceCatalogBuilder {
    /// Initializes a new instance of ResourceCatalogBuilder.
    pub fn new(id: ResourceCatalogId) -> Self {
        ResourceCatalogBuilder {
            id,
            properties: None,
            resources: None,
        }
    }

    /// Adds a property.
    pub fn with_property(&mut self, key: String, value: String) -> &mut Self {
        self.properties
            .get_or_insert_with(HashMap::new)
            .insert(key, value);

        self
    }

    /// Adds a readme.
    pub fn with_readme(&mut self, readme: String) -> &mut Self {
        self.with_property(README.to_string(), readme);
        self
    }

    /// Adds a license.
    pub fn with_license(&mut self, license: String) -> &mut Self {
        self.with_property(LICENSE.to_string(), license);
        self
    }

    /// Adds a resource.
    pub fn add_resource(&mut self, resource: Resource) -> &mut Self {
        self.resources.get_or_insert_with(Vec::new).push(resource);
        self
    }

    /// Adds a list of resources.
    pub fn add_resources(&mut self, resources: Vec<Resource>) -> &mut Self {
        self.resources
            .get_or_insert_with(Vec::new)
            .extend(resources);

        self
    }

    /// Builds the resource catalog.
    pub fn build(self) -> Result<ResourceCatalog, ResourcesError> {
        let resources = self.resources.map(Resources::try_new).transpose()?;

        Ok(ResourceCatalog {
            id: self.id,
            properties: self.properties,
            resources,
        })
    }
}
